pupil_src/shared_modules/video_capture/file_backend.py

```diff
@@ -220,6 +172,142 @@ def run_func(self, *args, **kwargs):
+        except av.AVError:
+            logger.info(
+                "Video at {0} is broken".format(self.containers_path[index]))
+            # return the next container
+            self.current_container_index += 1
+            return self._get_containers(self.current_container_index)
+        except IndexError:
+            raise NoMoreVideoError("No more container found")
+        else:
+            return container
+
+    def _set_timestamps(self, source_path):
+        '''
+        concatenate timestamp files (actual timestamps)
+        calculate np.diff and find indeces that represent gaps bigger than X
+        interpolate timestamps with frequence F and insert them into each gap
+        '''
+        self.timestamps_lst = self._get_timestamps_pattern_lst(
+            source_path)
+        avg_rate = self.video_stream.average_rate
+        if avg_rate is None:
+            avg_rate = Fraction(0, 1)
+
+        if float(avg_rate) % 1 != 0.0:
+            logger.error(
+                "Videofile pts are not evenly spaced, pts to" +
+                "index conversion may fail and be inconsitent."
+            )
+        self.timestamps = np.array([])
+        try:
+            for ts_filename in self.timestamps_lst:
+                self.frame_count = np.append(
+                    self.frame_count, len(np.load(ts_filename)))
+                self.timestamps = np.append(
+                    self.timestamps, np.load(ts_filename))
+        except IOError:
+            logger.warning(
+                "did not find timestamps file, making timetamps" +
+                "up based on fps and frame count. Frame count and" +
+                "timestamps are not accurate!"
+            )
+            # TODO: create self.timestamps from sketch
+            frame_rate = float(avg_rate)
+            self.timestamps = np.array([
+                i / frame_rate
+                for i in range(
+                    int(self.container[0].duration
+                        / av.time_base * frame_rate) + 100
+                )
+            ])  # we are adding some slack.
+        else:
+            logger.debug(
+                "Auto loaded %s timestamps from %s"
+                % (len(self.timestamps), self.timestamps_lst)
+            )
+        assert isinstance(
+            self.timestamps[0], float
+        ), "Timestamps need to be instances of python float, got {}".format(
+            type(self.timestamps[0])
+        )
+        ori_timestamps = self.timestamps.copy()
+        self.timestamps = self._interpolate_timestamps(self.timestamps)
+        self.timestamps_mask = np.in1d(self.timestamps, ori_timestamps)
+
+    def _interpolate_timestamps(self, timestamps):
+        first = np.load(self.timestamps_lst[0])
+        frame_rate = (first[-1] - first[0]) / len(first)
+        for i, t in np.ndenumerate(timestamps[:-1]):
+            if timestamps[i[0]+1] - timestamps[i[0]] > 0.5:
+                timestamps = np.append(
```

Comment:
`np.append()` is very slow. I would recommend using a list to aggregate timestamps:
```python
ts_aggregate = [timestamps]
for gap in gaps:
    filling_timestamps = interpolate_gap(gap)
    ts_aggregate.append(filling_timestamps)
timestamps = np.concatenate(ts_aggregate)
timestamps.sort()
```